### PR TITLE
Handle Dynamic Bounds in Shape Function of DynamicUpdateSliceOp

### LINKED ISSUE #799: Handle bounds in `DynamicUpdateSliceOp`shape function

### PR DESCRIPTION
**Pull request by sdasgup3**
The PR fixes #799 (please refer this for more details)

 Adding @smit-hinsu for review.

### CODE REVIEW

**Review by smit-hinsu — APPROVED** [review]
Thanks!

File: stablehlo/dialect/TypeInference.cpp [hunk 1]
```diff
@@ -1337,22 +1337,19 @@ LogicalResult inferDynamicUpdateSliceOp(
   auto operandType = operand.getType().cast<ShapedType>();
   auto updateType = update.getType().cast<ShapedType>();
 
-  // (C3)
```
Comment (ghpvnist) [inline stablehlo/dialect/TypeInference.cpp:1340]: Any reason for removing the constraints comment?

Comment (sdasgup3) [reply stablehlo/dialect/TypeInference.cpp]: Good catch!
I thought them to be stray comments. But realised that they are linked 1-1 to the spec constraints. Added them back.

File: stablehlo/dialect/TypeInference.cpp [hunk 2]
```diff
@@ -1388,8 +1388,9 @@ LogicalResult inferDynamicUpdateSliceOp(
 
   // (C1)
   if (operandType.hasRank()) {
-    inferredReturnShapes.emplace_back(operandType.getShape(),
-                                      operandType.getElementType());
+    inferredReturnShapes.emplace_back(
+        operandType.getShape(), operandType.getElementType(),
+        operandType.cast<RankedTensorType>().getEncoding());
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1393]: What are our plans for updating the rest of the code? More specifically, what's going to happen to C6?

Comment (zhouxin913) [reply stablehlo/dialect/TypeInference.cpp:1393]: I think this leads to a more general question: what do we do for an existing "X <= Y" check where X is a dim from one tensor, and Y is a dim from a different tensor, in the new bounds scenario? 
I would propose a util function below to replace:
```
assert(X <= Y);
```
with 
```
assert(LessEqualThanWithBounds(X, Y, A, B);
```
```
//  c4:  ?              any             true
//  c5:  any            ?               true
//  c0:  X              Y               assert X <= Y
//  c2:  X              ?, B            assert X <= B
//  c2:  ?, A           Y               true(?)
//  c6:  ?, B           ?, C            true(?)
LogicalResult LessEqualThanWithBounds(
    Optional<Location> location, int64_t dim, 
    int64_t leftSize, int64_t rightSize, 
    int64_t leftBound, int64_t rightBound) {
  // TODO
}
```

Comment (sdasgup3) [reply stablehlo/dialect/TypeInference.cpp:1393]: Thanks for summarizing. 
Added a ticket  to track this.

**Review by burmako — APPROVED** [review]

**Review by zhouxin913 — APPROVED** [review]

**Pull request merged**